pennylane/templates/subroutines/uccsd.py

```diff
@@ -183,10 +189,23 @@ def __init__(self, weights, wires, s_wires=None, d_wires=None, init_state=None,
                     f"expected entries of d_wires to be of size 2; got {d_wires_} of length {len(d_wires_)}"
                 )
 
+        if n_repeats < 1:
+            raise ValueError(f"Requires n_repeats to be at least 1; got {n_repeats}.")
+
         shape = qml.math.shape(weights)
-        if shape != (len(s_wires) + len(d_wires),):
+
+        expected_shape_1 = (len(s_wires) + len(d_wires),)
+        expected_shape_repeats = (n_repeats,) + expected_shape_1
+        if n_repeats == 1 and (
+            (len(shape) == 2 and shape != expected_shape_repeats)
+            or (len(shape) == 1 and shape != expected_shape_1)
+        ):
+            raise ValueError(
+                f"Weights tensor must be of shape {expected_shape_repeats} or {expected_shape_1}; got {shape}."
+            )
+        if n_repeats != 1 and shape != expected_shape_repeats:
```

Comment:
To cover all necessary cases I came up with the following:

```python
expected_shape_1 = (len(s_wires) + len(d_wires),)
expected_shape_repeats = (n_repeats,) + expected_shape_1
if len(shape) == 2:
    if shape != expected_shape_repeats:
        raise ValueError(
            f"Weights tensor must be of shape {expected_shape_repeats}; got {shape}."
        )
elif len(shape) == 1:
    if n_repeats != 1 or shape != expected_shape_1:
        raise ValueError(
            f"Weights tensor must be of shape {expected_shape_1 if n_repeats == 1 else expected_shape_repeats}; got {shape}."
        )
else:
    raise ValueError(
        f"Weights tensor must be of shape {expected_shape_repeats}; got {shape}."
    )
```

This is more readable but results in more lines of code. What are your thoughts @obliviateandsurrender?

Reply:
Thanks Eric, it looks good! It could be further simplified as follows: 

```python
expected_shape = (len(s_wires) + len(d_wires),)
if len(shape) == 1 and (n_repeats != 1 or shape != expected_shape):
    raise ValueError(
        f"For one-dimensional weights tensor, the shape must be {expected_shape}", and n_repeats should be 1; "
        f"got {shape} and {n_repeats}, respectively."
    )
if len(shape) != 1 and shape != (n_repeats,) + expected_shape:
    raise ValueError(
        f"Weights tensor must be of shape {(n_repeats,) + expected_shape}; got {shape}."
    )
```